pkg/util/ioutil/ioutil.go

```diff
@@ -7,69 +7,70 @@ package ioutil
 import (
 	"context"
 	"io"
+	"sync/atomic"
 	"time"
 )
 
 // timeoutReader monitors the progress of the io.Reader Read method
-// and calls the cancel function when reading does not progress for
+// and calls the cancelFn function when reading does not progress for
 // the specified time.
 type timeoutReader struct {
-	t *time.Timer
-	r io.Reader
-	n chan int
+	total  atomic.Uint64 // Total number of bytes read until timeout.
+	reader io.Reader
+
+	timer   *time.Timer
+	timeout time.Duration
+
+	done   atomic.Bool
+	cancel context.CancelFunc
 }
 
 // Read implements the io.Reader Read interface.
 func (tr *timeoutReader) Read(p []byte) (int, error) {
-	n, err := tr.r.Read(p)
-
-	v := n
-	if err != nil && tr.t.Stop() {
-		select {
-		case <-tr.t.C:
-		default:
-		}
-
-		// Negative value signals that error has
-		// occurred and the goroutine should terminate.
-		v = -1
-	}
-
-	select {
-	case tr.n <- v:
-	default:
+	n, err := tr.reader.Read(p)
+	switch {
+	case err != nil && !tr.done.Load():
+		tr.cancel()
+		tr.done.Store(true)
+	case !tr.done.Load():
+		tr.total.Add(uint64(n))
+		tr.timer.Reset(tr.timeout)
 	}
-
 	return n, err
 }
 
 // TimeoutReader creates a new timeoutReader instance and starts
 // a goroutine that monitors the progress of reading from the given reader.
 // If no progress is made for the duration of the given timeout, then the
-// reader executes the given cancel function and terminates the underlying
+// reader executes the given cancelFn function and terminates the underlying
 // goroutine. This goroutine will also terminate when the given context
 // is canceled or if an error is returned by the given reader.
-func TimeoutReader(ctx context.Context, r io.Reader, timeout time.Duration, cancel func(uint64)) io.Reader {
-	tr := &timeoutReader{t: time.NewTimer(timeout), r: r, n: make(chan int)}
+func TimeoutReader(ctx context.Context, reader io.Reader, timeout time.Duration, cancelFn func(uint64)) io.Reader {
+	ctx, cancel := context.WithCancel(ctx)
+	tr := &timeoutReader{
+		reader:  reader,
+		timer:   time.NewTimer(timeout),
+		timeout: timeout,
+		cancel:  cancel,
+	}
 
 	go func() {
-		var total uint64 = 0
+		defer func() {
+			tr.done.Store(true)
+			tr.timer.Stop()
+			select {
+			case <-tr.timer.C:
+			default:
+			}
+		}()
 
-		for {
+		for !tr.done.Load() {
```

Comment:
is condition in for loop nessesery? is for loop also nessesery? 

goroutine  will block on select statement and return once data is received on ether channel (there is no next iteration)